### PR TITLE
feat: template expressions

### PR DESCRIPTION
**Pull request by mattrym**
This PR introduces basic templating capabilities into `spc`. It allows transforming strings (passed in as parameters) and other, explicitly provided values according to some functions. 

A list of functions supported can be found [here](https://github.com/semaphoreci/spc/pull/47/files#diff-74688adaf42ca4fedf6e584eb1902b12c0a8766bf2bba2ea316034b1ae6acf6fR206). Implementation of these functions rely on [sprout](https://github.com/go-sprout/sprout) library, which is a successor of a well-established [sprig](https://github.com/Masterminds/sprig) library (used i.a. in [Helm](https://github.com/helm/helm/blob/main/go.mod#L9) or [Replicated](https://github.com/replicatedhq/kots/blob/main/go.mod#L12)). Functions were selected to allow basic operation without direct access to the running environment.

Syntax of templating expressions introduces two types of expressions:
* `${{ expression }}` produces a string as a result of template expression. This allows us to maintain backward compatibility with previous `${{ parameters.PARAM_NAME }}` syntax. 
* `%{{ expression }}` produces a structure which is JSON-serialized and injected directly in the pipeline YAML. In case that expression is surrounded by non-whitespace characters, is is printed as a JSON structure.

Both expressions maintain possibility to convert pipeline parameters using `parameters.PARAM_NAME` syntax.

Please check [test file](https://github.com/semaphoreci/spc/pull/47/files#diff-f44d648dc4d204e3bc2efc4df15f742bcfaea575df9863b43a536e0b7e99958b) for more examples of a usage.

### CODE REVIEW

File: pkg/pipelines/parameters_evaluator.go [hunk 1]
```diff
@@ -43,46 +44,188 @@ func (e *parametersEvaluator) Run() error {
...
-				strconv.Itoa(secretIndex),
-				"name",
-			})
+		e.extractFromSecretsAt(block.Search("task"), []string{"blocks", strconv.Itoa(blockIndex), "task"})
+	}
+}
+
+func (e *parametersEvaluator) ExtractFromContainerSecrets() {
+	for blockIndex := range e.pipeline.Blocks() {
+		e.tryExtractingFromPath([]string{"blocks", strconv.Itoa(blockIndex), "name"})
```
Comment (DamjanBecirovic) [inline pkg/pipelines/parameters_evaluator.go:85]: The function mentions container secrets, but it is processing block names. One or the other seems to be wrong.

File: pkg/pipelines/parameters_evaluator.go [hunk 2]
```diff
@@ -4,16 +4,17 @@ import (
 	"fmt"
 	"strconv"
 
+	"github.com/Jeffail/gabs/v2"
 	consolelogger "github.com/semaphoreci/spc/pkg/consolelogger"
-	parameters "github.com/semaphoreci/spc/pkg/parameters"
+	templates "github.com/semaphoreci/spc/pkg/templates"
```
Comment (DamjanBecirovic) [inline pkg/pipelines/parameters_evaluator.go:9]: With this change, we are no longer using the parameters package, so let's remove it. 
Also, we can then rename `parameteres_evaluator` into `templates_evaluator` to maintain clarity and we should also do the same change [here](https://github.com/semaphoreci/spc/blob/master/pkg/cli/compile.go#L23-L31).

File: pkg/templates/expression.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	consolelogger.Infof("Its value will be injected verbatim in the YAML file.\n")
+	consolelogger.EmptyLine()
+
+	exp.Value = expressionValue
+	return nil
+}
+
+func (exp *Expression) replaceValueInParsedString(expressionValue interface{}, regexMatch string) error {
+	if exprValueAsString, isString := expressionValue.(string); isString {
+		consolelogger.Infof("Expression produces a string as an.\n")
```
Comment (DamjanBecirovic) [inline pkg/templates/expression.go:150]: This log message is incomplete.

**Comment by DamjanBecirovic** [comment]
Can you also add a few template examples to yaml used for testing [here](https://github.com/semaphoreci/spc/blob/master/test/e2e/parameters_and_change_in.rb)?
It would be good to have a full e2e test and we can also examine what is logged there directly.

**Review by DamjanBecirovic — APPROVED** [review]

**Comment by TomFern** [comment]
Tracking doc updates in this task: https://github.com/semaphoreci/semaphore/issues/323

**Pull request merged**